tv_apps/browser/js/bookmarkList.js

```diff
@@ -0,0 +1,839 @@
+      }
+    },
+
+    /**
+     * Remove the focus list item.
+     */
+    removeFocusItem: function() {
+      var targetEl = null,
+          i = 0;
+
+      targetEl = this.listItemMap[this.focusIndex];
+      if(targetEl) {
+        this.resetItem(targetEl);
+        this.initializedItems.push(targetEl);
+
+        //reindex and update transform of the element behind the focus index
+        for(i = this.focusIndex; i < this.listIndexEndAt; i++) {
+          this.listItemMap[i] = this.listItemMap[i+1];
+          targetEl = this.listItemMap[i];
+          targetEl.setAttribute('data-index', i);
+          this.shiftItemTransform(targetEl, (-1 * LIST_ITEM_HEIGHT));
+        }
+
+        delete this.listItemMap[this.listIndexEndAt];
+        this.listIndexEndAt--;
+
+        // if the removed item is the last item of index,
+        // focus the new last element
+        if(this.listItemMap[this.focusIndex]) {
+          this.focusItem(this.listItemMap[this.focusIndex]);
+        } else if(this.listIndexEndAt != -1) {
+          this.focusIndex = this.listIndexEndAt;
+          this.focusItem(this.listItemMap[this.focusIndex]);
+        }
+
+        var event = new Event('itemRemoved');
+        this.el.dispatchEvent(event);
+      }
+    },
+
+    /**
+     * Focus list item.
+     * @param  {Node} el - HTML list item to focus.
+     */
+    focusItem: function(el) {
+      el.focus();
+      var event = new CustomEvent('focusItem', {detail: el});
+      this.el.dispatchEvent(event);
+    },
+
+    /**
+     * Focus the First visible list item.
+     */
+    focusFirstVisibleItem: function() {
+      var targetEl = null;
+
+      // if there is list item rendered and the current active element not in
+      // bookmark list, focus the first visible list item.
+      if(this.listIndexEndAt != -1 &&
+        Array.prototype.indexOf.call(
+          this.listEl.children,
+          document.activeElement) == -1) {
+        this.focusIndex = this.listVisibleStartAt;
+        targetEl = this.listItemMap[this.focusIndex];
+        if(targetEl) {
+          this.focusItem(targetEl);
+        }
+      }
+    },
+
+    /**
+     * Reset list item.
+     * @param  {Node} el - HTML list item to focus.
+     */
+    resetItem: function(el) {
+      var iconImgEl = el.querySelector('.icon-box-left .icon'),
+          titleEl = el.querySelector('.title'),
+          uriEl = el.querySelector('.uri');
+
+      el.setAttribute('data-index', -1);
+      el.setAttribute('data-type', '');
+      el.setAttribute('data-folder', '');
+      el.setAttribute('readOnly', false);
+      el.style.transform = '';
+      iconImgEl.style.backgroundImage = '';
+      iconImgEl.setAttribute('data-icon', '');
+      titleEl.textContent = '';
+      uriEl.textContent = '';
+    },
+
+    /**
+     * Preload the item before and after the view item.
+     * @param  {Number} index - current focus index.
+     */
+    loadItemBuffer: function(index) {
+      var bufferIndex = 0;
```

Comment:
I'm a bit worried of 2 things, 1) what if the data querying takes longer than expected. 2) preloading data in both ways seems not ideal. We can go with this now.